Lib/idlelib/pyshell.py

```diff
@@ -1221,32 +1285,35 @@ def recall(self, s, event):
             if prefix.rstrip().endswith(':'):
                 self.newline_and_indent_event(event)
                 prefix = self.text.get("insert linestart", "insert")
-            self.text.insert("insert", lines[0].strip())
+            self.text.insert("insert", lines[0].strip(),
+                             self.user_input_insert_tags)
             if len(lines) > 1:
                 orig_base_indent = re.search(r'^([ \t]*)', lines[0]).group(0)
                 new_base_indent  = re.search(r'^([ \t]*)', prefix).group(0)
                 for line in lines[1:]:
                     if line.startswith(orig_base_indent):
                         # replace orig base indentation with new indentation
                         line = new_base_indent + line[len(orig_base_indent):]
-                    self.text.insert('insert', '\n'+line.rstrip())
+                    self.text.insert('insert', '\n' + line.rstrip(),
+                                     self.user_input_insert_tags)
         finally:
             self.text.see("insert")
             self.text.undo_block_stop()
 
     def runit(self):
+        index_before = self.text.index("end-2c")
         line = self.text.get("iomark", "end-1c")
         # Strip off last newline and surrounding whitespace.
         # (To allow you to hit return twice to end a statement.)
-        i = len(line)
-        while i > 0 and line[i-1] in " \t":
-            i = i-1
-        if i > 0 and line[i-1] == "\n":
-            i = i-1
-        while i > 0 and line[i-1] in " \t":
-            i = i-1
-        line = line[:i]
-        self.interp.runsource(line)
+        line = re.sub(r"[ \t]*\n?[ \t]*$", "", line)
```

Comment:
I don't think that this, apparently untested, is equivalent to the no-unittest code it replaces.  This might be the source of the double blank line bug.  This sort of pure string manipulation should be in a separate function that can be unittested.

Reply:
I've reviewed this again just now and I'm rather sure that it is exactly equivalent, but if you're worried I can add some tests.

Reply:
It looks plausibly the same, but I did not see any other change that might plausibly cause the regression.  I will take another look and do some quick comparison tests without unittest boilerplate.

Reply:
Argh, you were right!

From [the re module's docs](https://docs.python.org/3/library/re.html#regular-expression-syntax):
> $ Matches the end of the string or just before the newline at the end of the string, [...]

Reply:
Fixed (with a rather excessive set of unit tests...)